src/profiler.ts

```diff
@@ -33,36 +40,267 @@ export interface ProfilerConfig extends AuthenticationConfig {
+
+    // TODO: enable heap profiling once heap profiler implemented.
+    this.profileTypes = [];
+    if (!this.config.disableTime) {
+      this.profileTypes.push(ProfileTypes.Wall);
+      this.timeProfiler =
+          new TimeProfiler(this.config.timeSamplingIntervalMicros);
+    }
   }
 
-  // Begins collection and uploading of profiles.
-  // If profiling fails or another problem is encountered, the returned promise
-  // will be rejected.
-  // TODO: implement
-  // TODO: explain failure situations.
+  /**
+   * Starts and endless loop to poll Stackdriver Profiler server for
+   * instructions, and collects and uploads profiles as requested.
+   * If there is a problem when collecting a profile or uploading a profile to
+   * Stackdriver Profiler, this problem will be logged at the debug level.
+   * If there is a problem polling Stackdriver Profiler for instructions
+   * on the type of profile created, this problem will be logged. If the problem
+   * indicates one definitely will not be able to profile, an error will be
+   * thrown.
+   */
   async start(): Promise<void> {
-    return Promise.reject(
-        new Error('start() is unimplemented for ProfileAgent.'));
+    return this.pollProfilerService();
+  }
+
+  /**
+   * Endlessly polls the profiler server for instructions, and collects and
+   * uploads profiles as requested.
+   */
+  async pollProfilerService(): Promise<void> {
+    const startCreateMillis = Date.now();
+    const prof = await this.createProfile();
+    await this.profileAndUpload(prof);
+    const endCreateMillis = Date.now();
+
+    // Schedule the next profile.
+    setTimeout(this.pollProfilerService.bind(this), 0).unref();
+  }
+
+  /**
+   * Talks to Stackdriver Profiler server, which hangs until server indicates
+   * job should be profiled.
+   *
+   * If any problem is encountered, the problem will be logged and
+   * createProfile() will be retried.
+   *
+   * TODO: implement backoff and retry when error encountered. createProfile()
+   * should be retried at time response indicates this request should be retried
+   * or with exponential backoff (up to one hour) if the response does not
+   * indicate when to retry this request. Once this is implemented, an error
+   * will be thrown only if the error indicates one definitely should not
+   * retry createProfile.
+   *
+   * Public to allow for testing.
+   */
+  async createProfile(): Promise<RequestProfile> {
+    const reqBody = {
+      deployment: {
+        projectId: this.config.projectId,
+        target: this.config.serviceContext.service,
+        labels: {zone: this.config.zone, instance: this.config.instance},
+      },
+      profileType: this.profileTypes,
+    };
+    const options = {
+      method: 'POST',
+      uri: '/profiles',
+      body: reqBody,
+      json: true,
+    };
+
+    let requestError: Error|undefined = undefined;
+    let retryRequest = true;
+    try {
+      const [body, response] = await this.request(options);
```

Comment:
`this.request` already includes `retry-request` logic. Is that not sufficient?

Reply:
Since we want to keep retrying indefinitely in certain cases, I don't think the logic in retry-request will be sufficient.

Reply:
In that case, I think you need higher level retry logic. Looking at HTTP response code and figuring out when to retry is duplicating logic from `retry-request`. Over time, this will drift apart from what `retry-logic` does.